logrotate.c

```diff
@@ -2451,15 +2451,51 @@ static int writeState(const char *stateFilename)
     strcpy(tmpFilename, stateFilename);
     strcat(tmpFilename, ".tmp");
     /* Remove possible tmp state file from previous run */
-    unlink(tmpFilename);
-
-
-    if ((fdcurr = open(stateFilename, O_RDONLY)) < 0) {
-        message(MESS_ERROR, "error opening %s: %s\n", stateFilename,
-                strerror(errno));
+    error = unlink(tmpFilename);
+    if (error == -1 && errno != ENOENT) {
+        message(MESS_ERROR, "error removing old temporary state file %s: %s\n",
+                tmpFilename, strerror(errno));
         free(tmpFilename);
         return 1;
     }
+    error = 0;
+
+    fdcurr = open(stateFilename, O_RDONLY);
+    if (fdcurr == -1) {
+        const char *state_header_v2 = "logrotate state -- version 2\n";
+        const size_t state_header_v2_len = strlen(state_header_v2);
+        ssize_t ret;
+
+        /* no error if state file is just missing */
+        if (errno != ENOENT) {
+            message(MESS_ERROR, "error opening state file %s: %s\n",
+                    stateFilename, strerror(errno));
+            free(tmpFilename);
+            return 1;
+        }
+
+        /* create a stub state file with mode 0640 */
+        fdcurr = open(stateFilename, O_CREAT | O_EXCL | O_WRONLY,
+                      S_IWUSR | S_IRUSR | S_IRGRP);
+        if (fdcurr == -1) {
+            message(MESS_ERROR, "error creating stub state file %s: %s\n",
+                    stateFilename, strerror(errno));
+            free(tmpFilename);
+            return 1;
+        }
+
+        /* check write access */
+        ret = write(fdcurr, state_header_v2, state_header_v2_len);
```

Comment:
Please use `full_write()` instead of `write()`.  `fopen/fprintf/fclose` is not really equivalent to `open/write/close`.

Reply:
changed, thanks

Reply:
changed, thanks